### PR TITLE
CIF - ServerError: storage.AccountsClient#ListAccountSAS: TooManyRequests

### PR DESCRIPTION
**Pull request by LiniSusan**
### Which issue this PR addresses:
https://issues.redhat.com/browse/ARO-3804

Fixes

### What this PR does / why we need it:
To report the cluster installation failure error related to storage.AccountsClient#ListAccountSAS: Failure responding to request: StatusCode=429

### CODE REVIEW

File: pkg/util/steps/runner.go [hunk 1]
```diff
@@ -54,6 +56,14 @@ func Run(ctx context.Context, log *logrus.Entry, pollInterval time.Duration, ste
 
 		if err != nil {
 			log.Errorf("step %s encountered error: %s", step, err.Error())
+			if strings.Contains(err.Error(), "ThrottlingLimitExceeded") {
+				err = api.NewCloudError(
+					http.StatusTooManyRequests,
+					api.CloudErrorCodeThrottlingLimitExceeded,
+					"",
+					"Requests are being throttled due to Azure Storage limits being exceeded. Please visit https://learn.microsoft.com/en-us/azure/openshift/troubleshoot for more details.")
```
Comment (swiencki) [inline pkg/util/steps/runner.go:64]: Can you update the link to this PR when it lands - https://github.com/MicrosoftDocs/azure-docs/pull/117572

File: pkg/util/storage/manager.go [hunk 2]
```diff
@@ -43,7 +46,23 @@ func (m *manager) BlobService(ctx context.Context, resourceGroup, account string
...
 	if err != nil {
+		if strings.Contains(err.Error(), "TooManyRequests") { 
+				err = &api.CloudError{
+					StatusCode: http.StatusTooManyRequests,
+					CloudErrorBody: &api.CloudErrorBody{
+						Code:    api.CloudErrorCodeThrottlingLimitExceeded,
+						Message: "ThrottlingLimitExceeded",
+						Details: []api.CloudErrorBody{
+							{
+								Message: string("Requests are being throttled due to Azure Storage limits being exceeded. Please visit https://learn.microsoft.com/en-us/azure/openshift/troubleshoot for more details."),
```
Comment (swiencki) [inline pkg/util/storage/manager.go:59]: Can you update the link to https://learn.microsoft.com/en-us/azure/openshift/troubleshoot#exceeding-azure-storage-limits

File: pkg/util/steps/runner.go (hunk 1 above)
Comment (LiniSusan) [reply pkg/util/steps/runner.go]: Updated

File: pkg/util/storage/manager.go (hunk 2 above)
Comment (LiniSusan) [reply pkg/util/storage/manager.go]: updated. Thanks @swiencki

File: pkg/util/steps/runner.go [hunk 3]
```diff
@@ -54,6 +57,16 @@ func Run(ctx context.Context, log *logrus.Entry, pollInterval time.Duration, ste
 
 		if err != nil {
 			log.Errorf("step %s encountered error: %s", step, err.Error())
+			if strings.Contains(err.Error(), "ThrottlingLimitExceeded") || strings.Contains(err.Error(), "TooManyRequests") {
+				correlation_id := fmt.Sprintf("%v", log.Data["correlation_id"])
+				message := fmt.Sprintf("Requests are being throttled due to Azure Storage limits being exceeded. Please visit https://learn.microsoft.com/en-us/azure/openshift/troubleshoot#exceeding-azure-storage-limits for more details. CorrelationId: " + correlation_id)
```
Comment (SrinivasAtmakuri) [inline pkg/util/steps/runner.go:62]:
```suggestion
				message := fmt.Sprintf("Requests are throttled due to Azure Storage limits being exceeded. Please visit https://learn.microsoft.com/en-us/azure/openshift/troubleshoot#exceeding-azure-storage-limits for more details. CorrelationId: " + correlation_id)
```

File: pkg/util/storage/manager.go [hunk 4]
```diff
@@ -43,7 +46,22 @@ func (m *manager) BlobService(ctx context.Context, resourceGroup, account string
...
 	if err != nil {
+		if strings.Contains(err.Error(), "TooManyRequests") {
+			err = &api.CloudError{
+				StatusCode: http.StatusTooManyRequests,
+				CloudErrorBody: &api.CloudErrorBody{
+					Code:    api.CloudErrorCodeThrottlingLimitExceeded,
+					Message: "ThrottlingLimitExceeded",
+					Details: []api.CloudErrorBody{
+						{
+							Message: string("Requests are being throttled due to Azure Storage limits being exceeded. Please visit https://learn.microsoft.com/en-us/azure/openshift/troubleshoot#exceeding-azure-storage-limits for more details."),
```
Comment (SrinivasAtmakuri) [inline pkg/util/storage/manager.go:59]: LGTM, a small nit,

```suggestion
							Message: string("Requests are throttled due to Azure Storage limits being exceeded. Please visit https://learn.microsoft.com/en-us/azure/openshift/troubleshoot#exceeding-azure-storage-limits for more details."),
```

File: pkg/util/steps/runner.go (hunk 3 above)
Comment (swiencki) [reply pkg/util/steps/runner.go:62]: Any details on why the error is happening should be within the link. I think this is too wordy

**Review by rhamitarora — APPROVED** [review]
LGTM

File: pkg/util/steps/runner.go (hunk 3 above)
Comment (jaitaiwan) [reply pkg/util/steps/runner.go:62]: `Requests are being throttled` is colloquial english. So that was fine wording. To shorten it you might be able to do something like `Requests Throttled. Visit () for more info. CorrelationId: `

**Review by jaitaiwan — APPROVED** [review]
I'm fine with longer errors, I prefer verbosity unless we can fit more information in less characters. Gave a suggestion for a shorter error message to satisfy Simon's request. LGTM!

File: pkg/util/storage/manager.go [hunk 5]
```diff
@@ -44,6 +47,20 @@ func (m *manager) BlobService(ctx context.Context, resourceGroup, account string
 		SharedAccessExpiryTime: &date.Time{Time: t.Add(24 * time.Hour)},
 	})
 	if err != nil {
+		if strings.Contains(err.Error(), "TooManyRequests") {
```
Comment (mociarain) [inline pkg/util/storage/manager.go:50]: Why match on a sting in the error message and not the `res.Response.StatusCode`?

Comment (LiniSusan) [reply pkg/util/storage/manager.go:50]: I dont think there is much impact when it is changed to `if err.(autorest.DetailedError).StatusCode == "TooManyRequests"`

Comment (mociarain) [reply pkg/util/storage/manager.go:50]: I was thinking more about matching against the number. Like we do here: https://github.com/azure/ARO-RP/blob/7cbf95b811d79677594809f90c8e45681a382b49/pkg/database/cosmosdb/zz_generated_cosmosdb.go#L73

This way if they change "TooManyRequest" -> "Too Many Requests" or whatever we're still good. IMO Matching on the statusCode is more robust. 

What do you think?

File: pkg/util/storage/manager.go (hunk 4 above)
Comment (LiniSusan) [reply pkg/util/storage/manager.go:58]: We generated this error message with reference to the original message that we go from Azure aside : "The request is being throttled as the limit has been reached for operation type - Read_ObservationWindow_00:05:00. For more information, see - https://aka.ms/srpthrottlinglimits"

File: pkg/util/steps/runner.go (hunk 3 above)
Comment (LiniSusan) [reply pkg/util/steps/runner.go]: We generated this error message with reference to the original message that we go from Azure errors : "The request is being throttled as the limit has been reached for operation type - Read_ObservationWindow_00:05:00. For more information, see - https://aka.ms/srpthrottlinglimits"

File: pkg/util/storage/manager.go (hunk 5 above)
Comment (LiniSusan) [reply pkg/util/storage/manager.go]: @mociarain : could you please review the code now? 
let me know if this is more convincing or not

**Review by mociarain — APPROVED** [review]
LGTM :)

File: pkg/util/storage/manager.go (hunk 5 above)
Comment (mociarain) [reply pkg/util/storage/manager.go]: That's exactly what I had in mind

File: pkg/util/storage/manager.go [hunk 6]
```diff
@@ -44,6 +46,22 @@ func (m *manager) BlobService(ctx context.Context, resourceGroup, account string
...
+						Message: "ThrottlingLimitExceeded",
+						Details: []api.CloudErrorBody{
+							{
+								Message: string("Requests are being throttled due to Azure Storage limits being exceeded. Please visit https://learn.microsoft.com/en-us/azure/openshift/troubleshoot#exceeding-azure-storage-limits for more details."),
+							},
+						},
+					},
+				}
+			}
+		}
```
Comment (AldoFusterTurpin) [inline pkg/util/storage/manager.go:64]: I think we could extract that to an auxiliar function and leverage the concept of early return to simplify the levels of nesting of the block of code. It is pretty common in Go to exit as soon as possible and I think we can take advantage of it 🙂 
Line 65 could become 
```go
return nil, getCorrectErrWhenTooManyRequests(err)
```
and getCorrectErrWhenTooManyRequests would be something like
```go
func getCorrectErrWhenTooManyRequests(err error) error {
	detailedError, ok := err.(autorest.DetailedError)
	if !ok {
		return err
	}

	if detailedError.StatusCode != http.StatusTooManyRequests {
		return err
	}

	msg := "Requests are being throttled due to Azure Storage limits being exceeded. Please visit https://learn.microsoft.com/en-us/azure/openshift/troubleshoot#exceeding-azure-storage-limits for more details."
	cloudError := &api.CloudError{
		StatusCode: http.StatusTooManyRequests,
		CloudErrorBody: &api.CloudErrorBody{
			Code:    api.CloudErrorCodeThrottlingLimitExceeded,
			Message: "ThrottlingLimitExceeded",
			Details: []api.CloudErrorBody{
				{
					Message: string(msg),
				},
			},
		},
	}

	return cloudError
}
```

**Review by AldoFusterTurpin — DISMISSED** [review]
@LiniSusan  That's good. Thank you for this. 
I left a suggestion to leverage early returns to make the code a bit more Go-ish, reducing the cyclomatic complexity.
🙂

File: pkg/util/storage/manager.go (hunk 6 above)
Comment (LiniSusan) [reply pkg/util/storage/manager.go]: updated the code with the mentioned changes..

**Comment by rhamitarora** [comment]
LGTM

**Review by rhamitarora — APPROVED** [review]

File: pkg/util/storage/manager.go [hunk 7]
```diff
@@ -33,6 +35,32 @@ func NewManager(env env.Core, subscriptionID string, authorizer autorest.Authori
...
 
+func getCorrectErrWhenTooManyRequests(err error) error {
+	detailedError, ok := err.(autorest.DetailedError)
+	if !ok {
+		return err
+	}
+
+	if detailedError.StatusCode != http.StatusTooManyRequests {
+		return err
+	} else {
```
Comment (AldoFusterTurpin) [inline pkg/util/storage/manager.go:46]: One last comment before LGTM, the else is not necessary due to the previous return (we can avoid one nested level then). One of the benefits of having a separate function for this 🙂

**Review by AldoFusterTurpin — CHANGES_REQUESTED** [review]
One last minor thing and then LGTM.

File: pkg/util/storage/manager.go (hunk 7 above)
Comment (AldoFusterTurpin) [inline pkg/util/storage/manager.go:46]: We can get rid of this else but just a minor thing.

**Review by AldoFusterTurpin — APPROVED** [review]
LGTM. One minor suggestions but definitely not a blocker 🙂

**Comment by swiencki** [comment]
Looks good :+1:

**Review by jaitaiwan — APPROVED** [review]
LGTM

**Review by SrinivasAtmakuri — APPROVED** [review]

**Review by Shivkumar13 — APPROVED** [review]
LGTM

**Comment by Shivkumar13** [comment]
Waiting for CI to go green, post that good to merge.

**Review by mociarain — APPROVED** [review]
